Add disconnection share and ACU recall options

lL1l1 opened the pull request:
It can be game-ruining to lose players because of disconnects, blowing up all the engineers in their base or units at their front and destroying all their units in share until death games. Especially because of recent connection issues/inherent Supcom connection issues. This PR fixes that with two new settings:
- DC Share Conditions: Determines how units are shared in case of a disconnect. To prevent abuse in Assassination, one ACU for that player must be at full HP and shields for the disconnect share conditions to apply. Default: Same as share condition
Common use case: share until death players could use fullshare for unfortunate disconnects
- Recall Disconnected ACUs: Instead of exploding, an ACU can recall and leave everything intact. To prevent abuse, the ACU must be at full HP and shields. This works individually on all ACUs a player owns. Default: Off
Common use case: could be used in team matchmaker to prevent games from being completely lost due to an early disconnect.

The full HP/shields threshold can be adjusted rather easily. Another idea is to not have enemy combat units within a large range. This would really ensure that the ACU is out of danger and the player isn't just abusing the feature.
Maybe the Recall Disconnected ACUs feature could use a time limit after which the rule goes away?
There is also an idea to leave the ACU around on disconnect, and apply fullshare rules until the ACU is killed, and applying the share condition rules then.

lL1l1 committed:
Add text for on-disconnect lobby options

lL1l1 committed:
Reorganize SimUtils.lua

lL1l1 committed:
Move army death functions from aibrain to simutils

lL1l1 committed:
Use newer version of DisableAI from OnDefeat

instead of from OnRecalled

lL1l1 committed:
Implement disconnect lobby options

lL1l1 committed:
upvalue functions

Garanas commented:
What if instead of checking whether the unit is full health, we instead check if the unit received damage in the last 10 seconds?

We could keep track (for all units) in what game tick that unit received damage for the last time.

lL1l1 commented:
Doesn't updating a value every time any unit takes damage just to check it once at the end of the game sound expensive? I'm not experienced, so this is more of a curiosity/learning question. I do agree it's better for ACUs, since regen/shields varies a lot between factions, and the feature shouldn't depend on faction or upgrade choice.

Garanas commented:
> Doesn't updating a value every time any unit takes damage just to check it once at the end of the game sound expensive?

Yes, as long as the complexity is O(1) then we're fine. There's a lot that happens when units take damage (health adjustments, veterancy) and that runs fine at the moment.

We could also use the same value to 'enhance' the CTRL + K ability - if a unit did not receive damage the past 10 seconds then it can be destroyed without triggering the death weapon for example.

Note that the '10 seconds' is arbitrarily chosen, can be any value.

apollodeathstar commented:
I feel the easiest choice is to just have the ACU be handed over along with the other units, if it was going to die it would still die anyway + more likely due the the "stall" that happens to units during the handover. This prevents abuse and a 99% health ACU getting shot by a single enemy tank on DC exploding and leading to the same outcome as now (which we are trying to avoid)

I would say either hand the ACU over or make the ACU neutral (civilian) and give it a "recall" CD timer of 30 seconds to 1 min after which it recalls off the field. (despawns)

lL1l1 commented:
> I feel the easiest choice is to just have the ACU be handed over along with the other units

> they are still being punished for the DC'ed players DC if they do not get it via the loss in combat power.

In a lot of maps and at increasing rating the shared economy and units can be more beneficial than the control or commentary from another player, enough to overcome the loss of an ACU. It has also been brought up on Discord that sharing ACUs would be extremely powerful (two gun coms perfectly coordinated).

> make the ACU neutral (civilian)

This relies on the map having a civilian army, and the fallback is killing (exploding) the ACU according to the civilian desertion share condition.

> Imagine if you did not get half the Factories or mexs from your allies because they DC'ed. The ACU early game is a small army onto itself.

This does remind me that the ACU has a large resource storage in the beginning of the game.

How about when a player disconnects, share their ACU + army and apply the share condition 5 minutes after the game started and 2 minutes after disconnect, or if the disconnected ACU dies early? The 5 minutes after start sharing allows use of the initial resources.  The 2 minutes after disconnect lets the ACU explode due to natural game rules instead of using an ACU recall option, and gives time to move units away, use the ACU one last time, and get tech to stabilize vs enemy ACU.
I think this would be the most balanced way to avoid abuse while preserving benefits. In assassination fullshare, which is the matchmaker mode, it would be most balanced. In assassination partial/no share, it wouldn't be fair because of the unit loss, but those are custom games that can rehost.
Could add it as an option instead of the ACU recall setting. Maybe like: "ACU Share: Explode, If-safe Recall, Delayed Recall, Permanent". One of them can be set as ranked if it's considered worthy. Permanent share might as well unlock giving your ACU to allies too.

lL1l1 committed:
Track ACUs last taking damage for disconnect rules

apollodeathstar commented:
I think having it share for something like 2 mins after disconnect is a good idea, after which it should just recall or explode. that is enough time for the player to compensate for the missing teammate and adjust their play. 

All I really care about doing is avoiding punishment for players in the early game who have a ally drop/dc and them suffering badly for it.  

Also the 2 mins timer means players will be unlikely to waste mass upgrading the ACU. I think that is a good way to handle it.

lL1l1 commented:
> We could also use the same value to 'enhance' the CTRL + K ability - if a unit did not receive damage the past 10 seconds then it can be destroyed without triggering the death weapon for example.

I think this is too overpowered for reclaiming volatile buildings, but it could be nice to prevent (possibly accidental) griefing with ACUs: if the ACU hasn't take damage recently and there are no enemies near the explosion range, it wouldn't detonate. After all, there's 0 benefit to the ACU dying without dealing damage because it leaves no wreckage.

lL1l1 committed:
Add ACU Sharing

lL1l1 commented:
The ACU sharing mechanic is functional but it still needs proper UI with LOC tags and a better "received units from [player]" message that tells you the time remaining for delayed recall and the share condition after disconnect for delayed recall/permanent ACU share.

lL1l1 committed:
Improve annotations

lL1l1 committed:
Implement sending a reason for units being shared

Refactor parameter for transfer of unfinished units

clyfordv commented:
Could:
1. Put a timer on the transferred ACU, after which it becomes un-selectable.
2. Similarly, start a timer when the receiving player gives the ACU an order, then make it un-selectable after that timer expires.
3. As an addendum/combined with the above, give the ACU a move order back to its spawn point when it's transferred.

lL1l1 commented:
> Put a timer on the transferred ACU

I did that. I'm writing a new announcement box so that it can 2 lines of text but I'm open to better ideas to show how long that ACU has left. Maybe a new tab or a timer in the avatars tab.

> after which it becomes un-selectable

Too out of the ordinary for how multiplayer skirmishes work, and it takes away control from the player for a key unit. It does make me wonder how the select/zoom to ACU hotkeys will work/break with multiple ACUs.

lL1l1 committed:
Add outside 8-way positioning methods

lL1l1 committed:
Refactor announcement.lua using layouter

lL1l1 committed:
Get rid of goalControl == textGroup case

it is impossible because GoalControl cannot be nil due to its creation in the beginning of the function if it is nil

lL1l1 committed:
Implement ItemList:SetAlphaOfColors

lL1l1 committed:
Implement `TextArea:FitToText()` and its helpers

Helpers: `TextArea:GetTextWidth()` and `TextArea:GetTextHeight()`

lL1l1 committed:
Implement `ItemList:GetAllItems`

lL1l1 committed:
Implement `AlignText` function

lL1l1 committed:
Implement TextArea SetTextAlignment

lL1l1 committed:
Make color.lua enum support case insensitive

lL1l1 committed:
Make new color functions enum support case insensitive

lL1l1 committed:
Optimize imports of LazyVar in UI classes

lL1l1 committed:
Fix typo in a uiutil comment

lL1l1 committed:
Implement using TextArea for announcements

lL1l1 committed:
Refactor "text" variable to "title" in announcement

lL1l1 committed:
Use share reasons in the share announcement

lL1l1 committed:
Add guards for share option announcement text

lL1l1 committed:
Final iteration of lobby options text

lL1l1 committed:
Use a variable for 2 the minute acu safety timer

lL1l1 committed:
Guard against abandoning without an ACU

could happen in non-assassination modes

lL1l1 committed:
Update a comment

lL1l1 committed:
Guard against disconnect abuse in ACU sharing

Applies the standard share condition if the ACU dies within the safety timer after being shared

lL1l1 committed:
Don't delayed recall ACUs in danger

prevents abusing the recall timer to, for example, not explode armies while fighting

lL1l1 committed:
First draft of changelog snippet

Too wordy? Are the examples needed? Needs images/video.

lL1l1 committed:
Fix unknown variable error

lL1l1 committed:
Fix sharing ACU in partial share

lL1l1 committed:
Fix concating nil when transferring 0 units

lL1l1 committed:
Simplify TransferUnitsToKiller

Also fix transferring 0 units to self

lL1l1 committed:
Allow `noRestrictions` to transfer to defeated brains

Fixes GetBackUnits(), specifically for shared units in traitor mode, compared to the base develop branch.

lL1l1 committed:
Delete announcementNew.lua

lL1l1 committed:
Fix typo in annotation comment

lL1l1 committed:
Init LastTickDamaged to 0 instead of checking for nil

Fixes KillArmyOnDelayedRecall function's LastTickDamaged check

lL1l1 committed:
Fix KillArmyOnDelayedRecall damage check logic

lL1l1 committed:
Kill ACUs when needed to prevent sharing

Explode -> all acus die
Recall/Recall Delayed -> unsafe acus die

lL1l1 committed:
Handle recalling ACUs

don't stack recalls on them
also don't count them as an ACU that keeps your army alive

lL1l1 committed:
Specify invalid share condition

lL1l1 committed:
Improve `TransferUnitsToHighestBrain`

- Don't put a `ratings` field into AIBrains when transferring units
- Give negative rated players some room
- Don't count AI rating when transferring

lL1l1 committed:
Annotate `CMauiItemList:SetNewColors`

lL1l1 committed:
Fix AutoLobbyConnectionMatrixDot

lL1l1 committed:
Update LaunchFAInstances.ps1

lL1l1 committed:
Fix in SimUtils TransferUnitsToHighestBrain usage

lL1l1 committed:
Remove brain.index

lL1l1 committed:
Merge branch 'develop' into feature/recall-on-disconnect

lL1l1 committed:
Merge branch 'develop' into feature/recall-on-disconnect

Garanas commented on lua/sim/units/ACUUnit.lua:
```diff
@@ -59,6 +62,19 @@ ACUUnit = ClassUnit(CommandUnit) {
+    ---@param damageType DamageType
+    OnDamage = function(self, instigator, amount, vector, damageType)
+        if self.CanTakeDamage and damageType ~= "TreeForce" and damageType ~= "TreeFire" then
+            self.LastTickDamaged = GetGameTick()
```
Won't this reference to `LastTickDamaged` be shared among all classes inheriting from this class? Because the value exists at the scope of the meta table. Not at the scope of the instance.

lL1l1 replied:
It won't because LastTickDamaged is a number and not a table. Tables would get shared at the meta table scope because the class system doesn't deep copy them, as we ran into when discussing enhancements.

Garanas commented on lua/ui/game/announcement.lua:
```diff
@@ -125,62 +165,49 @@ function CreateAnnouncement(text, goalControl, secondaryText, onFinished)
+    Layouter(bm):CenteredBelow(parent):FillHorizontally(parent):End()
+    Layouter(br):BottomRightOf(parent):End()
+
+    return { tl, tm, tr, ml, mr, bl, bm, br }
```
This is a different type of object then the original. It changes the interface. The original has hash-based references. Your implementation has index-based references. Are you sure that you can make this change without breaking UI mods?

lL1l1 replied:
I didn't notice I changed the type, I think I just wanted to minimize table operations by creating the table as the last step. I'll change it to the original hash format because this function has been global in the file since GPG so anyone could've imported it just to create its specific border.

Garanas reviewed (commented):
Some quick notes when purely reviewing the code.

lL1l1 committed:
Fix aibrain RecallArmyOnDefeat

lL1l1 committed:
Fix intellisense warning

The same file is required with different names.Lua Diagnostics.(different-requires)
simInit.lua(310, 12):

lL1l1 committed:
Fix the return type of CreateBorder

lL1l1 commented on lua/SimUtils.lua:
```diff
@@ -62,15 +34,18 @@ local function TransferUnitsOwnershipDelayedWeapons(weapon)
+function TransferUnitsOwnership(units, toArmy, captured, noRestrictions)
     local toBrain = GetArmyBrain(toArmy)
-    if not toBrain or toBrain:IsDefeated() or not units or table.empty(units) then
+    if not toBrain or (not noRestrictions and toBrain:IsDefeated()) or not units or table.empty(units) then
```
How about multiple lines like this?
```lua
if not toBrain or (toBrain:IsDefeated() and not noRestrictions)
    or table.empty(units)
then
    return
end
```

lL1l1 replied:
Done in 6baa71b

lL1l1 committed:
Fix runtime errors/intellisense warnings

lL1l1 committed:
Update ChangeUnitArmy annotation

lL1l1 committed:
Fix the enhancement disabling script

apparently I didn't notice the scope issue with `activeEnhancements`

lL1l1 committed:
Fix recall conditional

lL1l1 committed:
Revert "Fix recall conditional"

This reverts commit 1ee4618987a09898c4cc50823b608c2da998a59c.

lL1l1 committed:
Reuse GetGameTick call

lL1l1 committed:
Fix commander safe time logic at the start of game

lL1l1 committed:
Disable new options for future discussion

The refactor part of the PR is very useful, but the options are subjective, so they're disabled so that we can merge in the refactor changes without affecting gameplay.

lL1l1 committed:
Merge branch 'develop' into feature/recall-on-disconnect

lL1l1 committed:
Re-add sharedUnits table after merge

lL1l1 committed:
Fix merge mistake

lL1l1 committed:
Fix annotation warnings in SimUtils.lua

Except some `Army` type warnings and undefined global warnings for global functions in same file

lL1l1 committed:
Apply formatting to SimUtils.lua

lL1l1 committed:
Re-add decapitation's TransferToKiller handling

lL1l1 committed:
Clarify if statement

lL1l1 committed:
Revert multiplayer test script game options

lL1l1 committed:
Add a comment

lL1l1 committed:
Remove defunct weapon disabling during transfer code

lL1l1 committed:
Fix color.lua `EnumColors` backwards compat

Use a separate table to have case-insensitive color enum support in color lua functions.

05a99e3ff5697b8abfbe3330557e3d1040caf22a was problematic since it would break anything that ever imported the colors.lua enum table (3 years old), for example the connection matrix dot which I fixed manually (shouldn't have been necessary, but I'm keeping that change because it was misusing the enums).

Garanas commented:
Together with @lL1l1 we decided to merge the technical aspect of this pull request. The new functionality is not merged in. It needs some further discussion about what the behavior should be!
